rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+use yarp_sys::{yp_parser_t, yp_parser_init, yp_parse, yp_node_destroy, yp_node_t, yp_comment_t, yp_diagnostic_t};
+
+use std::ptr::NonNull;
+
+/// A diagnostic message that came back from the parser.
+#[derive(Copy, Clone, Debug)]
+pub struct Diagnostic(*mut yp_diagnostic_t);
+
+impl Diagnostic {
+    /// Returns the message associated with the diagnostic.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the message is not able to be converted into a `CStr`.
+    /// 
+    #[must_use]
+    pub fn message(&self) -> String {
+        unsafe {
+            let message: *mut i8 = (*(self.0)).message as *mut i8;
+            CStr::from_ptr(message).to_str().unwrap().to_string()
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Copy, Clone, Debug)]
+pub struct Comment(*mut yp_comment_t);
+
+impl Comment {
+    /// Returns the text of the comment.
+    #[must_use]
+    pub fn text(&self) -> String {
+        unsafe {
+            let start: *mut u8 = (*(self.0)).start as *mut u8;
+            let end: *mut u8 = (*(self.0)).end as *mut u8;
+
+            let length = end as usize - start as usize;
+            String::from_raw_parts(start, length, length)
+        }
+    }
+}
+
+/// A couple of different concepts are represented as linked lists in the
+/// parser. This includes errors, warnings, and comments. We want to be able to
+/// iterate over them in a consistent fashion. As such, we define a trait that
+/// can be used to iterate over them.
+pub trait ListNode<T> {
+    /// Returns true if the node is null.
+    fn is_null(&self) -> bool;
+
+    /// Returns the next node in the list.
+    fn next(&self) -> T;
+}
+
+impl ListNode<Self> for Diagnostic {
+    fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    fn next(&self) -> Self {
+        Self(unsafe { (*(self.0)).node.next.cast::<yp_diagnostic_t>() })
+    }
+}
+
+impl ListNode<Self> for Comment {
+    fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    fn next(&self) -> Self {
+        Self(unsafe { (*(self.0)).node.next.cast::<yp_comment_t>() })
+    }
+}
+
+/// An iterator that can be used to iterate over a linked list.
+pub struct ListIter<T: ListNode<T> + Copy> {
+    current: T
+}
+
+impl<T: ListNode<T> + Copy> Iterator for ListIter<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            None
+        } else {
+            let current = self.current;
+            self.current = self.current.next();
+            Some(current)
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult {
+    source: CString,
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl ParseResult {
+    /// Returns the source string that was parsed.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the source string is not able to be converted into a `&str`.
+    #[must_use]
+    pub fn source(&self) -> String {
```

Comment:
I would expect this API to return a borrowed view of the source rather than allocating.